### PR TITLE
[SQL] [SPARK-6794] Use kryo-based SparkSqlSerializer for GeneralHashedRelation

### PR DESCRIPTION
**Pull request by vlyubin**
Benchmarking results: http://pastie.org/private/1dneo1mta5zpsw6gmsoeq

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -46,8 +71,10 @@ private[joins] final class GeneralHashedRelation(hashTable: JavaHashMap[Row, Com
  * A specialized [[HashedRelation]] that maps key into a single value. This implementation
  * assumes the key is unique.
  */
-private[joins] final class UniqueKeyHashedRelation(hashTable: JavaHashMap[Row, Row])
-  extends HashedRelation with Serializable {
+private[joins] final class UniqueKeyHashedRelation(var hashTable: JavaHashMap[Row, Row])
```
Comment (aarondav) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: make this private, don't want to expose a mutable field publicly

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -29,16 +31,39 @@ import org.apache.spark.util.collection.CompactBuffer
  */
 private[joins] sealed trait HashedRelation {
   def get(key: Row): CompactBuffer[Row]
+
+  def writeBytes(out: ObjectOutput, serialized: Array[Byte]): Unit = {
```
Comment (aarondav) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: make protected and add a comment that these are helper functions to implement externalizable

**Comment by aarondav** [comment]
LGTM
cc @marmbrus

**Comment by chenghao-intel** [comment]
I must be lost in some context, why do we need to serde `HashReleation`? It supposed to be loaded once from the partition data iterator? isn't?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 3]
```diff
@@ -55,6 +82,14 @@ private[joins] final class UniqueKeyHashedRelation(hashTable: JavaHashMap[Row, R
   }
 
   def getValue(key: Row): Row = hashTable.get(key)
+
+  override def writeExternal(out: ObjectOutput): Unit = {
+    writeBytes(out, SparkSqlSerializer.serialize(hashTable))
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: This is probably very huge for `SparkSqlSerializer.serialize(hashTable)`, as in `ShuffledHashJoin`, we will load the whole partition data into the `hashTable`, which mean we probably keep 2 copies of the data in the memory, will that the intention?

**Comment by aarondav** [comment]
Note that this is only used for BroadcastHashJoins, and this is the broadcasted table. It is thus expected to be relatively small (< 10 MB). For Spark users with default configuration, the broadcast is serialized using Java serialization, which it turns out is much slower than Kryo in this case (see the benchmark).

This turns out to be a pretty significant win for very short O(seconds) queries, where a large portion of time may be spent in performing the broadcast.

**Comment by chenghao-intel** [comment]
OK, I see, for `BroadcastHashJoin`, thanks for explanation.

**Comment by marmbrus** [comment]
Merging to master.  Thanks!

**Pull request closed without merging**